### ISSUE TITLE
Check for Tags are mandatory and within allowed keys.

### ISSUE THREAD

**Issue by codebased** [aws-cloudformation/cfn-lint#2542]
I am trying to validate Tags: 

Check tags are defined for resources - if not throw Error
check tags contains all mandatory tags - If not throw Error

I am using the following code to cover both types of tags: 

```
Tags: 
  Key: Value
  Key: Value

```
And 

```
Tags: 
 - Key: XYZ
   Value: PQR
```

However it is unable to show error when I am using against this pattern:

```
Tags: 
  Key: Value
  Key: Value
```

Area of concern: 

  **message = "Missing Tag {0} at {1}"
                    message = message.format(required_tag, '/'.join(map(str, all_tag[:-1])))
                    matches.append(RuleMatch(all_tag[:-1],message))**

I see no error and I can see that these lines are executing. How do I debug?
 
```py
"""
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: MIT-0
"""
from cfnlint.rules import CloudFormationLintRule
from cfnlint.rules import RuleMatch


class PropertiesTagsRequired(CloudFormationLintRule):
    """Check if Tags have required keys"""
    id = 'W9002'
    # shortdesc = 'Tags have correct key values'
    # description = 'Check Tags for resources'
    # tags = ['resources', 'tags']
    # id = 'W9001' # New Rule ID
    shortdesc = 'test' # A short description about the rule
    description = 'test' # (Longer) description about the rule
    source_url = 'test' # A url to the source of the rule, e.g. documentation, AWS Blog posts etc
    tags = ["mappings", "limits"]

    def match(self, cfn):
        """Check Tags for required keys"""
        matches = list()
        
        required_tags = ['DataClassification', 'BluePrint']

        all_tags = cfn.search_deep_keys('Tags')
        all_tags = [x for x in all_tags if x[0] == 'Resources']
        
      # First tag format
        for all_tag in all_tags:
            all_keys = [d for d in all_tag[-1]]
            for required_tag in required_tags:
                if required_tag not in all_keys:
                    **message = "Missing Tag {0} at {1}"
                    message = message.format(required_tag, '/'.join(map(str, all_tag[:-1])))
                    matches.append(RuleMatch(all_tag[:-1],message))**
    
      # Second tag format

        for all_tag in all_tags:
            all_keys = [d.get('Key') for d in all_tag[-1]]
            for required_tag in required_tags:
                if required_tag not in all_keys:
                    message = "Missing Tag {0} at {1}"
                    matches.append(
                        RuleMatch(
                            all_tag[:-1],
                            message.format(required_tag, '/'.join(map(str, all_tag[:-1])))))

       
            
        return matches

```

**Comment by kddejong** [comment]
I'm not sure this scenario is fully valid:
```
Tags: 
  Key: Value
  Key: Value
```

That's considered a duplicate Key as you have Key defined twice.   Did you mean?
```
Tags: 
  Key1: Value
  Key2: Value
```

**Comment by kddejong** [comment]
I would write this rule like this:

```python
"""
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: MIT-0
"""
from cfnlint.rules import CloudFormationLintRule
from cfnlint.rules import RuleMatch


class PropertiesTagsRequired(CloudFormationLintRule):
    """Check if Tags have required keys"""

    id = "W9002"
    # shortdesc = 'Tags have correct key values'
    # description = 'Check Tags for resources'
    # tags = ['resources', 'tags']
    # id = 'W9001' # New Rule ID
    shortdesc = "test"  # A short description about the rule
    description = "test"  # (Longer) description about the rule
    source_url = "test"  # A url to the source of the rule, e.g. documentation, AWS Blog posts etc
    tags = ["mappings", "limits"]

    def _validate_tags(self, l: list) -> bool:
        pass

    def match(self, cfn):
        """Check Tags for required keys"""
        matches = list()

        required_tags = ["DataClassification", "BluePrint"]

        all_tags = cfn.search_deep_keys("Tags")
        all_tags = [x for x in all_tags if x[0] == "Resources"]

        # First tag format
        for all_tag in all_tags:
            tag_value = all_tag[-1]
            all_keys = []
            if isinstance(tag_value, dict):
                all_keys = [d for d in tag_value]
            elif isinstance(tag_value, list):
                all_keys = [d.get("Key") for d in tag_value]
            
            for required_tag in required_tags:
                if required_tag not in all_keys:
                    message = f"Missing Tag {required_tag} at {'/'.join(map(str, all_tag[:-1]))}"
                    matches.append(RuleMatch(all_tag[:-1], message))

        return matches


```
